server/embed/serve.go

```diff
@@ -99,7 +101,15 @@ func (sctx *serveCtx) serve(
 	splitHttp bool,
 	gopts ...grpc.ServerOption) (err error) {
 	logger := defaultLog.New(io.Discard, "etcdhttp", 0)
-	<-s.ReadyNotify()
+
+	// Make sure serversC is closed even if we prematurely exit the function.
+	defer close(sctx.serversC)
+
+	select {
+	case <-s.ReadyNotify(): // wait for e.Server to join the cluster
+	case <-s.StopNotify(): // publish aborted from 'ErrStopped'
+		return errors.ErrStopped
+	}
```

Comment:
This change is safe. I suggest to only keep this change and related test cases in this PR.